src/features/vault.ts

```diff
@@ -0,0 +1,241 @@
+    ? 'ANSIBLE_VAULT_IDENTITY_LIST'
+    : await utilAnsibleCfg.scanAnsibleCfg(rootPath);
+
+  if (!keyInCfg) {
+    displayMissingIdentityError();
+    return;
+  }
+
+  // Extract `ansible-vault` password
+
+  console.log(`Getting vault keyfile from ${keyInCfg}`);
+  vscode.window.showInformationMessage(
+    `Getting vault keyfile from ${keyInCfg}`
+  );
+
+  const text = editor.document.getText(selection);
+
+  // Go encrypt / decrypt
+  if (!!text) {
+    const type = getInlineTextType(text);
+
+    if (type === 'plaintext') {
+      console.log('Encrypt selected text');
+      const vaultId: string | undefined = await askForVaultId(keyInCfg);
+      if (!vaultId) {
+        displayMissingIdentityError();
+        return;
+      }
+
+      const encryptedText = `!vault |\n${encryptInline(
+        text,
+        rootPath,
+        vaultId,
+        config
+      )}`;
+      editor.edit((editBuilder) => {
+        editBuilder.replace(
+          selection,
+          encryptedText.replace(
+            /\n/g,
+            `\n${' '.repeat(selection.start.character)}`
+          )
+        );
+      });
+    } else if (type === 'encrypted') {
+      console.log('Decrypt selected text');
+
+      const decryptedText = await decryptInline(text, rootPath, config);
+      editor.edit((editBuilder) => {
+        editBuilder.replace(selection, decryptedText);
+      });
+    }
+  } else {
+    let content = '';
+    await vscode.workspace.openTextDocument(doc.fileName).then((document) => {
+      content = document.getText();
+    });
+    const type = getTextType(content);
+
+    if (type === 'plaintext') {
+      console.log('Encrypt entire file');
+      const vaultId: string | undefined = await askForVaultId(keyInCfg);
+      if (!vaultId) {
+        displayMissingIdentityError();
+        return;
+      }
+      encryptFile(doc.fileName, rootPath, vaultId, config);
+      vscode.window.showInformationMessage(`File encrypted: '${doc.fileName}'`);
+    } else if (type === 'encrypted') {
+      console.log('Decrypt entire file');
+
+      await decryptFile(doc.fileName, rootPath, config);
+      vscode.window.showInformationMessage(`File decrypted: '${doc.fileName}'`);
+    }
+    vscode.commands.executeCommand('workbench.action.files.revert');
+  }
+};
+
+// Returns whether the selected text is encrypted or in plain text.
+const getInlineTextType = (text: string) => {
+  if (text.trim().startsWith('!vault |')) {
+    text = text.replace('!vault |', '');
+  }
+
+  return text.trim().startsWith('$ANSIBLE_VAULT;') ? 'encrypted' : 'plaintext';
+};
+
+// Returns wheter the file is encrypted or in plain text.
+const getTextType = (text: string) => {
+  return text.indexOf('$ANSIBLE_VAULT;') === 0 ? 'encrypted' : 'plaintext';
+};
+
+const encryptInline = async (
+  text: string,
+  rootPath: string | undefined,
+  vaultId: string,
+  config: vscode.WorkspaceConfiguration
+) => {
+  const tmpFilename = await tmp.tmpName();
```

Comment:
We may want to also make sure to have as strict ACL setup as possible. Ansible itself even uses SElinux (where available), not only regular UNIX permissions to secure such files better.